substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs

```diff
@@ -705,71 +709,82 @@ where
 	async fn submit_events<EventStream>(
 		&mut self,
 		startup_point: &StartupPoint<Block>,
-		mut stream: EventStream,
+		stream: EventStream,
 		sink: Subscription,
 		rx_stop: oneshot::Receiver<()>,
 	) -> Result<(), SubscriptionManagementError>
 	where
-		EventStream: Stream<Item = NotificationType<Block>> + Unpin,
+		EventStream: Stream<Item = NotificationType<Block>> + Unpin + Send,
 	{
-		let mut stream_item = stream.next();
-
-		// The stop event can be triggered by the chainHead logic when the pinned
-		// block guarantee cannot be hold. Or when the client is disconnected.
-		let connection_closed = sink.closed();
-		tokio::pin!(connection_closed);
-		let mut stop_event = futures_util::future::select(rx_stop, connection_closed);
-
-		while let Either::Left((Some(event), next_stop_event)) =
-			futures_util::future::select(stream_item, stop_event).await
-		{
-			let events = match event {
+		// make the stream abortable
+		let (stream, cancel_handle) = futures::stream::abortable(stream);
+
+		// create a channel to propagate error messages
+		let (tx_send, mut tx_receive) = futures::channel::mpsc::channel(1);
+		let mut events = move |event| {
+			match event {
 				NotificationType::InitialEvents(events) => Ok(events),
 				NotificationType::NewBlock(notification) =>
 					self.handle_import_blocks(notification, &startup_point),
 				NotificationType::Finalized(notification) =>
 					self.handle_finalized_blocks(notification, &startup_point),
 				NotificationType::MethodResponse(notification) => Ok(vec![notification]),
-			};
-
-			let events = match events {
-				Ok(events) => events,
-				Err(err) => {
-					debug!(
-						target: LOG_TARGET,
-						"[follow][id={:?}] Failed to handle stream notification {:?}",
-						self.sub_id,
-						err
-					);
-					_ = sink.send(&FollowEvent::<String>::Stop).await;
-					return Err(err)
-				},
-			};
+			}
+			.inspect_err(|e| {
+				debug!(
+					target: LOG_TARGET,
+					"[follow][id={:?}] Failed to handle stream notification {:?}",
+					&self.sub_id,
+					e
+				);
+			})
+		};
 
-			for event in events {
-				if let Err(err) = sink.send(&event).await {
-					// Failed to submit event.
-					debug!(
-						target: LOG_TARGET,
-						"[follow][id={:?}] Failed to send event {:?}", self.sub_id, err
-					);
-
-					let _ = sink.send(&FollowEvent::<String>::Stop).await;
-					// No need to propagate this error further, the client disconnected.
-					return Ok(())
+		let stream = stream
```

Comment:
It's a bit tempting to add something similar to `pipe_from_stream` that takes a stream `Result<T, Error>` and bails on the first error similar to how `try_for_each` works to avoid having this channel to injected here just to get back the error.

Nothing that needs to be addressed in this PR but would be neat :)

Reply:
added to the pr